Implement standard SearchEngines

Dillon-Brown opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here unless your commit contains the issue number -->
<!--- Ensure that all code ``` is surrounded ``` by triple back quotes. This can also be done over multiple lines -->
- Implemented BasicSearchEngine
- Implemented LuceneSearchEngine
- Standardised SearchEngine output
- Many miscellaneous bug fixes and improvements.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
- All SearchEngines will now output the same results and come from a common core.
- I've left most of the original code for the Basic and AOD searching in ```modules/home```, there are perhaps 4-5 classes/files that can be totally removed from here but I've left them in for some degree of BC.
- In the interest of clean design I've separated ```UnifiedSearchAdvanced``` into two separate engines. Users will still be able to select the usual ```BasicAndAodEngine``` options and all current settings are still respected. How this works for a user using ```BasicAndAodEngine``` is that the SearchQuery will determine a likely default based on if the user has ```AOD``` enabled along with looking at the request data for older sugar searches. The front-end will then also display the old functionality of being able to switch between ```AOD``` and ```Basic``` based on what the users default has been set to.

All search pages (Elastic/AOD/Basic) now look like the following:

![SuiteSearch](https://user-images.githubusercontent.com/26431166/120240062-96084600-c257-11eb-960d-65c99443642e.png)

With multiple engines available, you will now be able to select directly on the results page:

![resultsEngine](https://user-images.githubusercontent.com/26431166/120764646-486a3280-c510-11eb-9454-ca31a99b6d72.png)

 
## How To Test This
<!--- Please describe in detail how to test your changes. -->
1. Test ElasticSearch
2. Test BasicSearch
3. Test LuceneSearch

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
<!--- Go over all the following points and check all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

<!--- Your pull request will be tested via Travis CI to automatically indicate that your changes do not prevent compilation. --->

<!--- If it reports back that there are problems, you can log into the Travis system and check the log report for your pull request to see what the problem was. --->

clemente-raposo commented on lib/Search/AOD/LuceneSearchEngine.php:
```diff
@@ -0,0 +1,178 @@
+        $cachePath = 'cache/modules/AOD_Index/QueryCache/' . md5($queryString);
+        if (is_file($cachePath)) {
+            $mTime = filemtime($cachePath);
+            if ($mTime > (time() - 5 * 60)) {
```
Hi @Dillon-Brown, does this mean we are caching query results for a given amount of time?

Dillon-Brown replied:
Yes, I've kept this functionality the same as before, only real change here is that I'm calling ```filemtime``` directly instead of using a method to calculate DST manually (Due to a PHP bug a decade ago).

clemente-raposo replied:
Hi @Dillon-Brown, thank you. I maybe wrong, though Ideally we should cache the query results? we may get results that are out of date right?

clemente-raposo replied:
@Dillon-Brown as you mentioned this is keeping existing behavior. This is something we can go over on a separate conversation

clemente-raposo commented on lib/Search/AOD/LuceneSearchEngine.php:
```diff
@@ -0,0 +1,178 @@
+    {
+        $searchResults = [];
+
+        foreach ($hits as $i => $array) {
```
Hi @Dillon-Brown, minor suggestion could we rename `$array` to `$hit`, makes it easier to understand, and with `$array` one may get confused with the reserved word. Anyway, just a sugestion

Dillon-Brown replied:
Agreed, pushed.

clemente-raposo replied:
thank you

clemente-raposo commented on lib/Search/AOD/LuceneSearchEngine.php:
```diff
@@ -0,0 +1,178 @@
+
+        foreach ($hits as $i => $array) {
+            $recordModule = $array->record_module;
+            $searchResults[$recordModule][$i] = $array->record_id;
```
Hi @Dillon-Brown, does the index in `$searchResults[$recordModule][$i]` matter? because using `$i` for the index may lead to gaps/jumps in the indexes. If we want incremental indexes without jumps we could just do`$searchResults[$recordModule][]  = $array->record_id;`

Dillon-Brown replied:
Nope it doesn't matter, good point. Pushed.

Dillon-Brown replied:
Nope it doesn't matter, good point. Pushed.

clemente-raposo replied:
thank you

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+
+        global $beanList, $current_user;
+
+        if (!empty($_REQUEST['advanced']) && $_REQUEST['advanced'] !== 'false') {
```
Hi @Dillon-Brown 

I imagine this was here before, though some questions:

- Is the code expecting to receive things from `$_REQUEST`?
- The engine is one of the possible implementations right? it can be this one or lucene or other?
- If we have a generic engine that can have several implementation, we probably should receive everything to the parameters. So that all engines receive it.

what do you think?

Dillon-Brown replied:
Removed, we won't be using request data here, this engine will always "BasicSearch" so it doesn't need to check for advanced.

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+
+                // Retrieve the original list view defs and store for processing in case of custom layout changes
+                require(__DIR__ . '/../../../modules/' . $seed->module_dir . '/metadata/listviewdefs.php');
+                $orig_listViewDefs = $listViewDefs;
```
Hi @Dillon-Brown probably this was here before, but we could we rename `$orig_listViewDefs` to keep with the standard, something like: `$originalListViewDefs` or similar?

Dillon-Brown replied:
Pushed

clemente-raposo replied:
thank you

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+                    }
+                }
+
+                if (!empty($where_clauses)) {
```
Hi @Dillon-Brown could we also rename `$where_clauses` to `$whereClauses` or similar please?

Dillon-Brown replied:
Pushed

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+
+                $moduleCounts[$moduleName] = $listData['pageData']['offsets']['total'];
+
+                foreach ($listData['data'] as $i => $array) {
```
Hi @Dillon-Brown if could also rename `$array` to `$hit` or `$entry` or `$result`, please?

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+    {
+        global $beanList, $beanFiles, $dictionary;
+
+        $supported_modules = [];
```
Hi @Dillon-Brown could we rename `$supported_modules` to `$supportedModules` to keep with standards please?

Dillon-Brown replied:
Pushed

clemente-raposo commented on lib/Search/BasicSearch/BasicSearchEngine.php:
```diff
@@ -0,0 +1,526 @@
+                }
+
+                if (!empty($fields)) {
+                    $supported_modules [$moduleName] ['fields'] = $fields;
```
Hi @Dillon-Brown minor formatting change needed on `$supported_modules [$moduleName] ['fields'] `.

Dillon-Brown replied:
Pushed, strange my IDE doesn't want to fix these though 🤔

clemente-raposo replied:
thank you

clemente-raposo commented on lib/Search/UI/templates/search.form.tpl:
```diff
@@ -44,54 +50,46 @@
-                    return true;
-                }
-            };
+        let searchForm = {
```
Hi @Dillon-Brown I'm not sure if its safe already to use `let` without a transpiler. Though it seems to be supported by most browsers in the latest versions:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let

Dillon-Brown replied:
Changed back to var, will need to look into this further at some point as my IDE wants to auto-replaces these to either ```let``` or ```const```.

clemente-raposo commented on lib/Search/SqlSearch/SimpleSqlSearchEngine.php:
```diff
@@ -83,16 +89,16 @@ public function search(SearchQuery $query)
             }
 
-            if (strpos($type, 'text') == 0) {
+            if (str_starts_with($type, 'text')) {
```
Hi @Dillon-Brown I think `str_starts_with` its only supported in php8

Dillon-Brown replied:
We have polyfills for these, I'm going to revert anyway though as I believe the last Suite-7 release we had some issues with the polyfills when it came to running the upgrade wizard. Nice spot.

clemente-raposo commented on lib/Search/SearchQuery.php:
```diff
@@ -154,39 +199,67 @@ private static function filterArray(array $array, $key, $default, $filter = null
+    {
+        global $sugar_config;
+
+        if (!empty($sugar_config['search']['defaultEngine'])) {
```
Hi @Dillon-Brown, what should this method return if `$sugar_config['search']['defaultEngine']` is empty

Dillon-Brown replied:
Changed to return ```BasicSearchEngine``` when the config is not set. I think that would make the most sense since it's the only engine that doesn't require any setup and works OOTB.

Dillon-Brown commented:
@clemente-raposo Pushed two additional changes.

1. Deprecated ```UnifiedSearchAdvanced``` as none of this code is now used (Kept for BC but would prefer to remove completely, thoughts?)
2. Removed PHP 7.2 from Travis config and replaced it with PHP 7.3. Also pointed the codecoverage to 7.3 for performance reasons.

Dillon-Brown commented:
@clemente-raposo Latest changes include a new way to select engines from the results page. This also works for any custom engines the user may have added:

![extendedEngines](https://user-images.githubusercontent.com/26431166/120766551-3f7a6080-c512-11eb-89cd-60ca311e22f7.png)

clemente-raposo reviewed (dismissed):
code looks ok

Dillon-Brown commented:
Latest changes are just rebasing onto latest develop, all conflicts due to https://github.com/salesagility/SuiteCRM/pull/9170 which is just PHPUnit changes.

clemente-raposo reviewed (approved):
latest changes look ok
